crates/rust-analyzer/src/conv.rs

```diff
@@ -334,6 +335,75 @@ impl ConvWith<&FoldConvCtx<'_>> for Fold {
     }
 }
 
+impl ConvWith<(&FileId, &WorldSnapshot, &LineIndex)> for Vec<StructureNode> {
+    type Output = lsp_types::DocumentSymbolResponse;
+
+    fn conv_with(self, ctx: (&FileId, &WorldSnapshot, &LineIndex)) -> Self::Output {
+        let file_id = ctx.0;
+        let world = ctx.1;
+        let line_index = ctx.2;
+        let supports_hierarchy = world.config.client_caps.hierarchical_symbols;
+
+        if supports_hierarchy {
+            let mut parents: Vec<(DocumentSymbol, Option<usize>)> = Vec::new();
+
+            for symbol in self.into_iter() {
+                let doc_symbol = DocumentSymbol {
+                    name: symbol.label,
+                    detail: symbol.detail,
+                    kind: symbol.kind.conv(),
+                    deprecated: Some(symbol.deprecated),
+                    range: symbol.node_range.conv_with(line_index),
+                    selection_range: symbol.navigation_range.conv_with(line_index),
+                    children: None,
+                };
+                parents.push((doc_symbol, symbol.parent));
+            }
+            let mut res = Vec::new();
+            while let Some((node, parent)) = parents.pop() {
+                match parent {
+                    None => res.push(node),
+                    Some(i) => {
+                        let children = &mut parents[i].0.children;
+                        if children.is_none() {
+                            *children = Some(Vec::new());
+                        }
+                        children.as_mut().unwrap().push(node);
+                    }
+                }
+            }
+            res.into()
+        } else {
```

Comment:
I think a better approach would be to do this on the LSP level -- write a (recursive) function which flattens a bunch of `DocumentSymbol`s into a falt list of `SymbolInformation`. 

My gut feeling that we should try to use this lsp-level dumbing down approach in as much cases as possible.